src/react-components/hub-create-panel.js

```diff
@@ -64,21 +83,28 @@ class HubCreatePanel extends Component {
     return new RegExp(HUB_NAME_PATTERN).test(this.state.name) && new RegExp(hubAlphaPattern).test(this.state.name);
   };
 
-  setToEnvironmentOffset = offset => {
+  _preloadImage = async src => {
+    const img = new Image();
+    const imgLoad = new Promise(resolve => img.addEventListener("load", resolve));
+    img.srcset = src;
+    await imgLoad;
+  };
+
+  setToEnvironmentOffset = async offset => {
     const numEnvs = this.props.environments.length;
 
-    this.setState(state => ({
-      environmentIndex: ((state.environmentIndex + offset) % this.props.environments.length + numEnvs) % numEnvs
-    }));
+    const environmentIndex = ((this.state.environmentIndex + offset) % numEnvs + numEnvs) % numEnvs;
+    const environmentThumbnail = this._getEnvironmentThumbnail(environmentIndex);
+    await this._preloadImage(environmentThumbnail.srcset);
```

Comment:
If I'm reading this right, this blocks any visible changes to the UI when the next/prev button is hit until the image loads? (All our thumbnails are the same now so can't readily test.) I'm not sure that's better than showing a placeholder and letting the image load in, since users may think something is broken.

Reply:
If we're worried about it, i'd suggest adding an async prefetcher to load the rest of the images optimistically once you flip `ready: true` -- that way most likely they'll be already loaded but if not the user will still be able to cycle the environments without being blocked if they're not fetched yet. (of course you can get even smarter by loading just the neighboring images, but seems overkill until we have a lot of choices.)

Reply:
Alright, changed it to optimistically load all thumbnails now. It delays displaying the hub creation form entriely. We'll have to get a sense for it with actual images in production.